docs/references/ingress-messages.mdx

```diff
@@ -0,0 +1,82 @@
+
+# Ingress messages on ICP
+
+<MarkdownChipRow labels={["Intermediate","Reference", "Security"]} />
+
+## Overview
+
+Ingress messages are calls to ICP canisters that are sent by external entities, such as users, usually through an agent like [agent-js](https://github.com/dfinity/agent-js/tree/main) that runs within a web application, or [agent-rs](https://github.com/dfinity/agent-rs/tree/main) in a Rust application. See also the introductory [call overview](/docs/current/developer-docs/smart-contracts/call/overview). 
+
+For example, you can browse what ingress messages can be sent to the ckBTC ledger by browsing through its [API on the dashboard](https://dashboard.internetcomputer.org/canister/mxzaz-hqaaa-aaaar-qaada-cai). The tool also allows to send messages, using agent-js behind the scenes.
+
+In the following, we discuss ICP's ingress message APIs. While these APIs are defined in detail in the [HTTPS interface specification](/docs/current/references/ic-interface-spec#http-interface), we provide a more high level and intuitive overview. In addition to discussing the happy path, we lay a special focus on error handling. That aspect is particularly important as it can be tricky to determine if an ingress message has actually been successfully executed. In the worst case, misinterpreting errors could lead to bugs such as double spending.
+
+## Types of ingress messages
+
+ICP supports several types of calls, as defined in the [HTTPS interface specification](/docs/current/references/ic-interface-spec#http-interface). These include query and update calls. 
+
+## Query Calls
+Queries don't modify a canister's state and are answered and signed by a single replica. Responses are obtained within milliseconds, but they are typically not signed by the subnet, which implies that a malicious replica or boundary node could arbitrarily modify the response. 
+
+Some applications may have higher authenticity requirements for query responses. For example, if a dapp queries a proposal description that the user then votes on, it may not be acceptable that a single replica or boundary node can tamper with that information, as they could trick users to vote in the attacker's favor. One can consider issuing query calls as update calls instead, see the next section. While that's easy to implement, it makes the calls significantly slower. To avoid this negative performance impact, for some use cases one can sign the data beforehand and provide [certified variables](/docs/current/developer-docs/security/security-best-practices/data-integrity-and-authenticity#certified-variables) in query responses. However, this can be complex to implement, depending on the specific use case. An example where certified variables are used is the [HTTP asset certification](/docs/current/developer-docs/security/security-best-practices/data-integrity-and-authenticity#use-http-asset-certification-and-avoid-serving-your-dapp-through-rawicp0io). 
```

Comment:
```suggestion
Some applications may have higher authenticity requirements for query responses. For example, if a dapp queries a proposal description that the user then votes on, it may not be acceptable that the result could be tampered with, as it could trick users to vote in an attacker's favor. As an alternative, you can issue query calls as update calls instead. Although implementing this alternative is easy, it is significantly slow.

Another alternative approach is to sign the data beforehand and provide [certified variables](/docs/current/developer-docs/security/security-best-practices/data-integrity-and-authenticity#certified-variables) in query responses. While this approach has better performance than sending query calls as update calls, it can be complex to implement depending on the specific use case. An example where certified variables are used is within the [HTTP asset certification library](/docs/current/developer-docs/security/security-best-practices/data-integrity-and-authenticity#use-http-asset-certification-and-avoid-serving-your-dapp-through-rawicp0io). 
```

Reply:
Very nice, thanks for streamlining this, it reads much better. 

Merged this with Ogi's input above.